Kuadrant/multi-cluster-traffic-controller
# Webhooks setup and server

## Description

Add the resources and logic to set up a webhook server running alongside the controller, as well as the reconciliation of the webhook configuration

## Verification steps

In order to verify these changes, the controller must run in-cluster and be exposed via an Ingress.

1. Run the local-setup
    ```sh
    make local-setup
    ```
1. Build the controller image
    ```sh
    # Build the image
    make docker-build
    # Load it into the cluster
    kind load docker-image controller:latest --name mctc-control-plane  --nodes mctc-control-plane-control-plane
    ```
1. Deploy the controller. Prior to that, ensure that the manager includes the required environment variables to run
    ```sh
    make deploy
    ```
1. Deploy the secret with the kubeconfig for the control cluster
1. Verify the controller Ingress is created
    ```sh
    kubectl get ingress mctc-ingress-mctc -n multi-cluster-traffic-controller-system
    ```
    ```yaml
    kind: Ingress
    apiVersion: networking.k8s.io/v1
    metadata:
      name: mctc-ingress-mctc
      namespace: multi-cluster-traffic-controller-system
      annotations:   
        mctc-component: webhook 
    spec:
      ingressClassName: nginx
      tls:
        - hosts:
            - cwxkxzmah79fqyecvplwzf.sf.hcpapps.net
          secretName: cwxkxzmah79fqyecvplwzf.sf.hcpapps.net
      rules:
        - host: host.whatever.com
          http:
            paths:
              - path: /
                pathType: Prefix
                backend:
                  service:
                    name: mctc-mctc-webhooks
                    port:
                      number: 8082
        - host: cwxkxzmah79fqyecvplwzf.sf.hcpapps.net
          http:
            paths:
              - path: /
                pathType: Prefix
                backend:
                  service:
                    name: mctc-mctc-webhooks
                    port:
                      number: 8082
    status:
      loadBalancer:
        ingress:
          - ip: 172.21.0.2
    ```
1. Verify that the webhook configuration is reconciled
    ```sh
     kubectl get mutatingwebhookconfiguration mctc -o yaml
    ```
    Ensure it references the webhook server and the `caClient` field is set
    ```yaml
    // ...
    clientConfig:
      caBundle: ...
      url: https://cwxkxzmah79fqyecvplwzf.sf.hcpapps.net/ingress
    // ...    
    ```
2. Verify that the webhook handler is reached successfully by making a change in the webhook ingress. For example, adding an annotation
    ```sh
    kubectl edit ingress mctc-ingress-mctc -n multi-cluster-traffic-controller-system
    ```
    The Pod logs should append `level=info msg="WEBHOOK HANDLER"`, and the update should be performed correctly
3. In order to verify independently that the webhook creates successfully the new DNSRecord and Certificate, disable the traffic controller
    1.  Comment out the code that starts up the controller at [main.go](https://github.com/Kuadrant/multi-cluster-traffic-controller/blob/c7293b4d6bc1d792c872a3f1e2694f5c5b0adbba/main.go#L119-L127)
    1. Follow step 2 to build the image
    2. Delete the controller Pod so that a new Pod with the new image is created
4. Create the sample service so that the creation of the Ingress triggers the webhook
    ```sh
    kubectl apply -f samples/echo-service/echo.yaml -n default
    ```
5. Verify that the DNSRecord and the Certificate are created
    ```sh
    kubectl get dnsrecords s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net -n argocd
    NAME                                    AGE
    s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net   117m

    kubectl get certificate s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net -n argocd
    NAME                                    READY   SECRET                                  AGE
    s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net   True    s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net   118m
    ```
6. Verify that the Ingress is mutated
    ```sh
    kubectl get ingress echo -n default -o yaml
    ```
    ```yaml
    apiVersion: networking.k8s.io/v1
    kind: Ingress
    metadata:
      annotations:
        kuadrant.io/managed-hosts: s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net
      creationTimestamp: "2023-01-25T10:07:15Z"
      generation: 1
      name: echo
      namespace: default
      resourceVersion: "8576"
      uid: caee7c05-cefe-43d6-bd67-6cd56a3033cd
    spec:
      ingressClassName: nginx
      rules:
      - http:
          paths:
          - backend:
              service:
                name: echo
                port:
                  number: 80
            path: /
            pathType: Prefix
      - host: s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net
        http:
          paths:
          - backend:
              service:
                name: echo
                port:
                  number: 80
            path: /
            pathType: Prefix
      tls:
      - hosts:
        - s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net
        secretName: s6yxqyfx3awf53iasbk7a6.sf.hcpapps.net
    status:
      loadBalancer:
        ingress:
        - ip: 172.21.0.2
    ```
7. 
    ![image](https://user-images.githubusercontent.com/12643415/214557666-7a7383c5-b62e-4cb2-8d32-0b5fa88c593a.png)
